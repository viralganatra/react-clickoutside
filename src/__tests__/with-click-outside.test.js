import React from 'react';
import withClickOutside from '../with-click-outside';
import EventName from '../events';

jest.mock('../events', () => 'mousedown');

const MockComponent = () => <div className="container" />;

describe('Click Outside Higher Order Component', () => {
    const globalAddEvent = global.addEventListener;
    const globalRemoveEvent = global.removeEventListener;

    let eventMap;
    let ComposedComponent;

    beforeEach(() => {
        eventMap = {};

        global.addEventListener = jest.fn((event, cb) => {
            event && (eventMap[event] = cb);
        });

        global.removeEventListener = jest.fn((event) => {
            event && (delete eventMap[event]);
        });

        ComposedComponent = withClickOutside()(MockComponent);
    });

    afterEach(() => {
        global.addEventListener = globalAddEvent;
        global.removeEventListener = globalRemoveEvent;
    });

    it('should detect the click event outside of the component', () => {
        const spy = jest.fn();

        const wrapper = mount(
            <div>
                <p className="title">Title</p>
                <ComposedComponent onClickOutside={spy} />
            </div>,
        );

        eventMap.mousedown({ target: wrapper.find('.container').instance() });

        expect(spy).not.toHaveBeenCalled();

        eventMap.mousedown({ target: wrapper.find('.title').instance() });

        expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should unregister the event listener when the component unmounts', () => {
        const wrapper = mount(<ComposedComponent onClickOutside={jest.fn()} />);

        wrapper.unmount();

        expect(global.removeEventListener.mock.calls[4][0]).toEqual(EventName);
    });

    it('should add any custom properties to the container div', () => {
        const ComponentWithProps = withClickOutside({ className: 'foo' })(MockComponent);

        const wrapper = shallow(<ComponentWithProps onClickOutside={jest.fn()} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('should throw an error if called incorrectly', () => {
        expect(() => withClickOutside()()).toThrowErrorMatchingSnapshot();
    });
});
