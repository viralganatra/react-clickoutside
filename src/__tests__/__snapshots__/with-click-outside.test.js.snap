// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Click Outside Higher Order Component should add any custom properties to the container div 1`] = `
<div
  className="foo"
>
  <MockComponent />
</div>
`;

exports[`Click Outside Higher Order Component should throw an error if called incorrectly 1`] = `
"Expected \\"WrappedComponent\\" provided as the first argument to ClickOutsideHOC
            to be a function. Instead, received undefined."
`;
